kernel: add macro to define threads to run in kernel mode only

dcpleung opened the pull request:
This adds a macro `K_KERNEL_THREAD_DEFINE()` to enable defining threads that only run in kernel mode (when USERSPACE is enabled).

The first commit was originally from https://github.com/zephyrproject-rtos/zephyr/pull/54948, and has been slightly modified. A test has also been added to test it such a thread cannot be made into a user thread.

albertofloyd committed:
kernel: add macro to define threads to run in kernel mode only

This addition allows to define threads without a privileged stack,
and are to run in kernel mode only.

When CONFIG_USER_SPACE is enabled, a thread ends up allocating
space for both its own thread stack and a privileged stack.
When a thread runs in kernel mode only, this extra space for
the privileged stack is not needed.

The macro K_THREAD_DEFINE(), by default, defines a thread with
privileged stack. A new macro K_KERNEL_THREAD_DEFINE() is added
so that a kernel mode only thread can be defined without
the privileged stack.

DL: Edited commit message, clarification in function comment and
    alignment.

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>
Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: kernel/userspace: add a test for kernel only thread

This adds a test to make sure kernel only threads cannot go
into user mode: doing so would result in kernel panic.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: extract common code in K_{,KERNEL_}THREAD_DEFINE

K_THREAD_DEFINE and K_KERNEL_THREAD_DEFINE have common code
so extract the common code and put them into another macro
Z_THREAD_COMMON_DEFINE().

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -732,6 +732,54 @@ struct _static_thread_data {
+ * @param options Thread options.
+ * @param delay Scheduling delay (in milliseconds), zero for no delay.
+ *
+ * @internal It has been observed that the x86 compiler by default aligns
+ * these _static_thread_data structures to 32-byte boundaries, thereby
+ * wasting space. To work around this, force a 4-byte alignment.
```
I think this internal comment should be removed as there is nothing at this level of implementation referencing a 4-byte alignment. The alignment is auto-magically taken care of via the STRUCT_SECTION_ITERABLE().

dcpleung replied:
Removed comment on both.

dcpleung committed:
kernel: remove outdated comment in K_THREAD_DEFINE()

The internal comment in K_THREAD_DEFINE() on alignment needed
for x86 is outdated. The alignment is taken care of by
the iterable sections in linker scripts. So remove.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
